File: backend/app/models/content.py
"""ไฟล์แนบ คลังความรู้ การแจ้งเตือน และร่องรอยการใช้งาน — 7 ตาราง

ตรงกับ docs/02-data-model.md v2.0 §2.5, §4.6, §4.7
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    BigInteger,
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.constants import (
    KbStatus,
    KbVisibility,
    NotificationChannelCode,
    NotificationStatus,
    ScanStatus,
    values,
)
from app.db.base import Base, SoftDeleteMixin, TimestampMixin


class Attachment(Base, SoftDeleteMixin):
    __tablename__ = "attachment"
    __table_args__ = (
        CheckConstraint(f"scan_status IN {tuple(values(ScanStatus))}", name="scan_status_valid"),
        CheckConstraint("file_size <= 20971520", name="file_size_max_20mb"),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    # ทั้งสามเป็น null ได้ชั่วคราว เพราะ POST /attachments อัปโหลด "ก่อน" สร้าง ticket (B-08)
    # งาน cleanup_orphan_attachments ลบไฟล์ที่ไม่ผูกกับอะไรเกิน 24 ชม.
    ticket_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("ticket.id"), index=True)
    comment_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("ticket_comment.id"))
    kb_article_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("kb_article.id"))

    # path สัมพัทธ์ {company_id}/{yyyy}/{mm}/{uuid}.{ext} — ระบบเป็นคนกำหนดทั้งหมด
    # ชื่อไฟล์ของผู้ใช้ไม่เคยถูกใช้ประกอบ path (กัน path traversal)
    storage_key: Mapped[str] = mapped_column(String(255), unique=True, nullable=False)
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    # ตรวจจาก magic bytes จริง ไม่เชื่อ Content-Type ที่ client ส่งมา (NFR-15)
    mime_type: Mapped[str] = mapped_column(String(100), nullable=False)
    file_size: Mapped[int] = mapped_column(BigInteger, nullable=False)
    uploaded_by: Mapped[int] = mapped_column(BigInteger, ForeignKey("app_user.id"), nullable=False)
    # รองรับ virus scan ในเฟส 2 — เฟส 1 เป็น skipped ทั้งหมด (B-04)
    scan_status: Mapped[str] = mapped_column(
        String(20), default=ScanStatus.SKIPPED.value, nullable=False
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class KbCategory(Base):
    __tablename__ = "kb_category"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    parent_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("kb_category.id"))
    name_th: Mapped[str] = mapped_column(String(150), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)


class KbArticle(Base, TimestampMixin, SoftDeleteMixin):
    __tablename__ = "kb_article"
    __table_args__ = (
        CheckConstraint(f"visibility IN {tuple(values(KbVisibility))}", name="visibility_valid"),
        CheckConstraint(f"status IN {tuple(values(KbStatus))}", name="status_valid"),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    kb_category_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("kb_category.id"), nullable=False
    )
    # null = ทุกบริษัทเห็น
    company_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("company.id"), index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    summary: Mapped[str | None] = mapped_column(String(500))
    body_markdown: Mapped[str] = mapped_column(Text, nullable=False)
    visibility: Mapped[str] = mapped_column(
        String(20), default=KbVisibility.PUBLIC.value, nullable=False
    )
    status: Mapped[str] = mapped_column(
        String(20), default=KbStatus.DRAFT.value, nullable=False
    )
    tags: Mapped[str | None] = mapped_column(String(255))
    author_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("app_user.id"), nullable=False)
    view_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    helpful_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    not_helpful_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    source_ticket_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("ticket.id"))
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class KbFeedback(Base):
    """กันโหวตซ้ำรายผู้ใช้ (B-05 / US-13 AC-3)"""

    __tablename__ = "kb_feedback"
    __table_args__ = (UniqueConstraint("kb_article_id", "user_id"),)

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    kb_article_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("kb_article.id", ondelete="CASCADE"), nullable=False
    )
    user_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("app_user.id"), nullable=False)
    is_helpful: Mapped[bool] = mapped_column(Boolean, nullable=False)
    note: Mapped[str | None] = mapped_column(String(500))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class Notification(Base):
    """1 แถว = 1 ช่องทาง เพื่อให้ retry รายช่องทางแยกกันได้"""

    __tablename__ = "notification"
    __table_args__ = (
        CheckConstraint(
            f"channel IN {tuple(values(NotificationChannelCode))}", name="channel_valid"
        ),
        CheckConstraint(f"status IN {tuple(values(NotificationStatus))}", name="status_valid"),
        Index("ix_notification_unread", "user_id", "channel", "read_at"),
        # กัน Celery retry สร้างแถวซ้ำในวันเดียวกัน (S-04)
        Index(
            "uq_notification_dedup",
            "user_id",
            "ticket_id",
            "event_type",
            "channel",
            text("(created_at::date)"),
            unique=True,
        ),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("app_user.id"), nullable=False)
    ticket_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("ticket.id"))
    event_type: Mapped[str] = mapped_column(String(40), nullable=False)
    channel: Mapped[str] = mapped_column(String(20), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), default=NotificationStatus.PENDING.value, nullable=False
    )
    retry_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    error_message: Mapped[str | None] = mapped_column(String(500))
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class NotificationChannel(Base, TimestampMixin):
    __tablename__ = "notification_channel"
    __table_args__ = (
        UniqueConstraint("user_id", "channel"),
        CheckConstraint(
            f"channel IN {tuple(values(NotificationChannelCode))}", name="channel_valid"
        ),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("app_user.id", ondelete="CASCADE"), nullable=False
    )
    channel: Mapped[str] = mapped_column(String(20), nullable=False)
    destination: Mapped[str | None] = mapped_column(String(255))
    is_enabled: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    # LINE ต้องผูกบัญชีสำเร็จก่อนจึงส่งได้
    is_verified: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)


class AuditLog(Base):
    """append-only — ห้าม UPDATE/DELETE จากชั้น application

    เก็บอย่างน้อย 1 ปี และห้าม purge ก่อน 90 วัน (NFR-18 / นโยบาย 3.3)
    """

    __tablename__ = "audit_log"
    __table_args__ = (
        Index("ix_audit_created", text("created_at DESC")),
        Index("ix_audit_actor", "actor_id", text("created_at DESC")),
        Index("ix_audit_entity", "entity_type", "entity_id"),
        Index("ix_audit_company", "company_id", text("created_at DESC")),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    actor_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("app_user.id"))
    company_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("company.id"))
    action: Mapped[str] = mapped_column(String(40), nullable=False)
    entity_type: Mapped[str] = mapped_column(String(40), nullable=False)
    entity_id: Mapped[int | None] = mapped_column(BigInteger)
    # เฉพาะฟิลด์ที่เปลี่ยน — ห้ามเก็บ password / token (NFR-19)
    old_value: Mapped[dict | None] = mapped_column(JSONB)
    new_value: Mapped[dict | None] = mapped_column(JSONB)
    ip_address: Mapped[str | None] = mapped_column(String(45))  # รองรับ IPv6
    user_agent: Mapped[str | None] = mapped_column(String(255))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

File: backend/src/db/schema/content.ts
/**
 * ไฟล์แนบ คลังความรู้ การแจ้งเตือน และร่องรอยการใช้งาน — 7 ตาราง
 * ตรงกับ docs/02-data-model.md v2.0 §2.5, §4.6, §4.7
 */

import { sql } from 'drizzle-orm';
import {
  bigint,
  bigserial,
  boolean,
  check,
  index,
  integer,
  jsonb,
  pgTable,
  text,
  timestamp,
  unique,
  uniqueIndex,
  varchar,
} from 'drizzle-orm/pg-core';

import {
  KB_STATUS,
  KB_VISIBILITY,
  NOTIFICATION_CHANNEL,
  NOTIFICATION_STATUS,
  SCAN_STATUS,
} from '../../common/constants';
import { appUser, company, inList } from './organization';
import { ticket, ticketComment } from './ticket';

const timestamps = {
  createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
  updatedAt: timestamp('updated_at', { withTimezone: true }).defaultNow().notNull(),
};

export const kbCategory = pgTable(
  'kb_category',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    parentId: bigint('parent_id', { mode: 'number' }),
    nameTh: varchar('name_th', { length: 150 }).notNull(),
    sortOrder: integer('sort_order').default(0).notNull(),
    isActive: boolean('is_active').default(true).notNull(),
  },
  // ชื่อซ้ำภายใต้แม่เดียวกันไม่ได้ — parent_id เป็น NULL สำหรับหมวดระดับบนสุด
  // ซึ่งเป็นกรณีที่ต้องกันซ้ำมากที่สุด จึงต้อง NULLS NOT DISTINCT
  (t) => [unique('uq_kb_category_parent_name').on(t.parentId, t.nameTh).nullsNotDistinct()],
);

export const kbArticle = pgTable(
  'kb_article',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    kbCategoryId: bigint('kb_category_id', { mode: 'number' })
      .notNull()
      .references(() => kbCategory.id),
    // null = ทุกบริษัทเห็น
    companyId: bigint('company_id', { mode: 'number' }).references(() => company.id),
    title: varchar('title', { length: 255 }).notNull(),
    summary: varchar('summary', { length: 500 }),
    bodyMarkdown: text('body_markdown').notNull(),
    visibility: varchar('visibility', { length: 20 }).default('public').notNull(),
    status: varchar('status', { length: 20 }).default('draft').notNull(),
    tags: varchar('tags', { length: 255 }),
    authorId: bigint('author_id', { mode: 'number' })
      .notNull()
      .references(() => appUser.id),
    viewCount: integer('view_count').default(0).notNull(),
    helpfulCount: integer('helpful_count').default(0).notNull(),
    notHelpfulCount: integer('not_helpful_count').default(0).notNull(),
    sourceTicketId: bigint('source_ticket_id', { mode: 'number' }).references(() => ticket.id),
    publishedAt: timestamp('published_at', { withTimezone: true }),
    deletedAt: timestamp('deleted_at', { withTimezone: true }),
    ...timestamps,
  },
  (t) => [
    check('ck_kb_visibility_valid', inList('visibility', KB_VISIBILITY)),
    check('ck_kb_status_valid', inList('status', KB_STATUS)),
    // เผยแพร่แล้วต้องมีเวลาเผยแพร่
    check('ck_kb_published_needs_date', sql`status <> 'published' OR published_at IS NOT NULL`),
    index('ix_kb_article_company').on(t.companyId),
    index('ix_kb_article_status').on(t.status, t.visibility),
  ],
);

export const attachment = pgTable(
  'attachment',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    // ทั้งสามเป็น null ได้ชั่วคราว เพราะ POST /attachments อัปโหลด "ก่อน" สร้าง ticket (B-08)
    // งาน cleanup_orphan_attachments ลบไฟล์ที่ไม่ผูกกับอะไรเกิน 24 ชม.
    ticketId: bigint('ticket_id', { mode: 'number' }).references(() => ticket.id),
    commentId: bigint('comment_id', { mode: 'number' }).references(() => ticketComment.id),
    kbArticleId: bigint('kb_article_id', { mode: 'number' }).references(() => kbArticle.id),

    // path สัมพัทธ์ {company_id}/{yyyy}/{mm}/{uuid}.{ext} — ระบบเป็นคนกำหนดทั้งหมด
    // ชื่อไฟล์ของผู้ใช้ไม่เคยถูกใช้ประกอบ path (กัน path traversal)
    storageKey: varchar('storage_key', { length: 255 }).notNull().unique(),
    fileName: varchar('file_name', { length: 255 }).notNull(),
    // ตรวจจาก magic bytes จริง ไม่เชื่อ Content-Type ที่ client ส่งมา (NFR-15)
    mimeType: varchar('mime_type', { length: 100 }).notNull(),
    fileSize: bigint('file_size', { mode: 'number' }).notNull(),
    uploadedBy: bigint('uploaded_by', { mode: 'number' })
      .notNull()
      .references(() => appUser.id),
    // รองรับ virus scan ในเฟส 2 — เฟส 1 เป็น skipped ทั้งหมด (B-04)
    scanStatus: varchar('scan_status', { length: 20 }).default('skipped').notNull(),
    deletedAt: timestamp('deleted_at', { withTimezone: true }),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
  },
  (t) => [
    check('ck_attachment_scan_status_valid', inList('scan_status', SCAN_STATUS)),
    check('ck_attachment_size_max_20mb', sql`file_size <= 20971520`),
    index('ix_attachment_ticket').on(t.ticketId),
  ],
);

/** กันโหวตซ้ำรายผู้ใช้ (B-05 / US-13 AC-3) */
export const kbFeedback = pgTable(
  'kb_feedback',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    kbArticleId: bigint('kb_article_id', { mode: 'number' })
      .notNull()
      .references(() => kbArticle.id, { onDelete: 'cascade' }),
    userId: bigint('user_id', { mode: 'number' })
      .notNull()
      .references(() => appUser.id),
    isHelpful: boolean('is_helpful').notNull(),
    note: varchar('note', { length: 500 }),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
  },
  (t) => [unique('uq_kb_feedback_article_user').on(t.kbArticleId, t.userId)],
);

/** 1 แถว = 1 ช่องทาง เพื่อให้ retry รายช่องทางแยกกันได้ */
export const notification = pgTable(
  'notification',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    userId: bigint('user_id', { mode: 'number' })
      .notNull()
      .references(() => appUser.id),
    ticketId: bigint('ticket_id', { mode: 'number' }).references(() => ticket.id),
    eventType: varchar('event_type', { length: 40 }).notNull(),
    channel: varchar('channel', { length: 20 }).notNull(),
    title: varchar('title', { length: 255 }).notNull(),
    body: text('body').notNull(),
    status: varchar('status', { length: 20 }).default('pending').notNull(),
    retryCount: integer('retry_count').default(0).notNull(),
    errorMessage: varchar('error_message', { length: 500 }),
    sentAt: timestamp('sent_at', { withTimezone: true }),
    readAt: timestamp('read_at', { withTimezone: true }),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
  },
  (t) => [
    check('ck_notification_channel_valid', inList('channel', NOTIFICATION_CHANNEL)),
    check('ck_notification_status_valid', inList('status', NOTIFICATION_STATUS)),
    index('ix_notification_unread').on(t.userId, t.channel, t.readAt),
    // กัน BullMQ retry สร้างแถวซ้ำในวันเดียวกัน (S-04)
    //
    // ต้องตรึงโซนเวลาในนิพจน์ ใช้ created_at::date เฉย ๆ ไม่ได้สองเหตุผล
    //   1) Postgres ปฏิเสธ เพราะการแปลง timestamptz เป็น date ขึ้นกับ TimeZone
    //      ของ session จึงเป็น STABLE ไม่ใช่ IMMUTABLE ซึ่งดัชนีต้องการ
    //   2) ต่อให้ยอมรับ ขอบเขต "วัน" ก็จะเลื่อนตามโซนเวลาของผู้เชื่อมต่อ
    //      งานที่รันจาก worker คนละโซนจะตัดวันไม่ตรงกัน
    // timezone(text, timestamptz) เป็น IMMUTABLE จึงใช้เป็นดัชนีได้
    uniqueIndex('uq_notification_dedup').on(
      t.userId,
      t.ticketId,
      t.eventType,
      t.channel,
      sql`((created_at AT TIME ZONE 'Asia/Vientiane')::date)`,
    ),
  ],
);

export const notificationChannel = pgTable(
  'notification_channel',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    userId: bigint('user_id', { mode: 'number' })
      .notNull()
      .references(() => appUser.id, { onDelete: 'cascade' }),
    channel: varchar('channel', { length: 20 }).notNull(),
    destination: varchar('destination', { length: 255 }),
    isEnabled: boolean('is_enabled').default(true).notNull(),
    // LINE ต้องผูกบัญชีสำเร็จก่อนจึงส่งได้
    isVerified: boolean('is_verified').default(false).notNull(),
    ...timestamps,
  },
  (t) => [
    unique('uq_notification_channel_user').on(t.userId, t.channel),
    check('ck_notification_channel_code_valid', inList('channel', NOTIFICATION_CHANNEL)),
  ],
);

/**
 * append-only — ห้าม UPDATE/DELETE จากชั้น application
 * เก็บอย่างน้อย 1 ปี และห้าม purge ก่อน 90 วัน (NFR-18 / นโยบาย 3.3)
 */
export const auditLog = pgTable(
  'audit_log',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    actorId: bigint('actor_id', { mode: 'number' }).references(() => appUser.id),
    companyId: bigint('company_id', { mode: 'number' }).references(() => company.id),
    action: varchar('action', { length: 40 }).notNull(),
    entityType: varchar('entity_type', { length: 40 }).notNull(),
    entityId: bigint('entity_id', { mode: 'number' }),
    // เฉพาะฟิลด์ที่เปลี่ยน — ห้ามเก็บ password / token (NFR-19)
    oldValue: jsonb('old_value'),
    newValue: jsonb('new_value'),
    ipAddress: varchar('ip_address', { length: 45 }), // รองรับ IPv6
    userAgent: varchar('user_agent', { length: 255 }),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
  },
  (t) => [
    index('ix_audit_created').on(t.createdAt),
    index('ix_audit_actor').on(t.actorId, t.createdAt),
    index('ix_audit_entity').on(t.entityType, t.entityId),
    index('ix_audit_company').on(t.companyId, t.createdAt),
  ],
);
